arch: i.MX add RDC peripheral permission setting for applications cores

diegosueiro opened the pull request:
This patch adds the RDC (Resource Domain Controller) peripheral
permissions settings for the i.MX applications cores (Cortex A9 on
i.MX6 and Cortex A7 on i.MX7).

This will enable both Linux (on application's core) and Zephyr (on M4
core) to share the peripherals and coexist.

A complete solution should involve the SEMA4 to control the peripherals
access and prevent resource deadlocking and misusage.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>

MaureenHelm commented on arch/arm/soc/nxp_imx/mcimx6x_m4/soc.c:
```diff
@@ -13,90 +13,77 @@
-	RDC_SetPdapAccess(RDC, rdcPdapUart1,
-			RDC_DOMAIN_PERM(CONFIG_DOMAIN_ID, RDC_DOMAIN_PERM_RW),
-			false, false);
+	RDC_SetPdapAccess(RDC, rdcPdapUart1, RDC_DOMAIN_M4_A9_PERM_RW, false,
```
I think we may want to consider configuring these attributes in dts instead. @galak, what do you think?

diegosueiro replied:
@MaureenHelm,

Good suggestion.
Do you have a similar implementation that I can use as a base?

MaureenHelm replied:
Take a look the Kinetis sim clock controller. Also, is there any precedent in the Linux kernel?

diegosueiro replied:
@MaureenHelm ,

Thanks for the tip. I'll take a look on that.

As far as I know, the Linux side is not doing anything with the RDC configuration. I'll double check at the u-boot level.

diegosueiro replied:
@MaureenHelm,

I've been looking at adding the RDC permissions configuration into devicetree and I want to know if I'm in the right direction.
This is what I think I'll have to do. Please, correct if I'm wrong:

1 - Create a include/dt-bindings/rdc/imx_rdc.h file defining the permissions and domain id values.
2 - In each peripheral dts binding nxp,imx_[gpio|uart|i2c|pwm].yaml file add the rdc property as an array with the first value as a domain id and the second as the permission.
3 - Create a scripts/dts/extract/rdc.py to extract and create the defines the for the domain id and permission.
4 - Set the default rdc values per peripheral in the nxp_[imx6sx|imx7d]_m4.dtsi files and potentially override in the board dts file.
5 - Create the appropriate defines in the board dts.fixup
6 - In the soc.c use the defines generated by the devicetree processing.

diegosueiro replied:
@MaureenHelm ,

I pushed another version of the patch and I think I found a clear solution.

diegosueiro committed:
arch: i.MX add RDC peripheral permission setting for applications cores

This patch adds the RDC (Resource Domain Controller) peripheral
permissions settings for the i.MX applications cores (Cortex A9 on
i.MX6 and Cortex A7 on i.MX7).

This will enable both Linux (on application's core) and Zephyr (on M4
core) to share the peripherals and coexist.

The settings are defined at devicetree level and applied in the soc.c.

A complete solution should involve the SEMA4 to control the peripherals
access and prevent resource deadlocking and misusage.

Signed-off-by: Diego Sueiro <diego.sueiro@gmail.com>
